### PR TITLE
Added support for NS autoloadbalancing

### PR DESCRIPTION
**Pull request by HaruChebrolu**
# Description

Changes to support NS autoloadbalancing
Doc: https://docs.google.com/document/d/1bpx21903Sof1JzGTg81lG7lnk4QeCcqo6uF28Au_zEc/edit#heading=h.m22j2pb1oqna
Polarion:
CEPH-83598716
CEPH-83598717

Logs link: http://magna002.ceph.redhat.com/cephci-jenkins/harika/cephci-run-YR9PMT/

<summary>click to expand checklist</summary>

- [x] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [x] Review the automation design
- [x] Implement the test script and perform test runs
- [x] Submit PR for code review and approve
- [x] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

**Review by rahullepakshi — COMMENTED** [review]
Can we avoid running "gw info" command so many times? It creates confusion while analyzing logs

Also namespace list or probably any CLI output is logged twice. Any idea why? can we avoid it?

File: tests/nvmeof/workflows/ha.py [hunk 1]
```diff
@@ -446,6 +460,127 @@ def power_on_off(self, gateway, action, wait_for_active_state=None):
...
+
+        Post scale-down Validation
+        - List out namespaces associated with the scaled down Gateways using ANA group ids.
+        - Check for 5 Consecutive times for the increments in write/read to validate IO continuation.
+        """
+        fail_gws, _ = self.catogorize(gateway_nodes)
+        start_counter = float()
+        start_time = str()
+        end_counter = float()
+        end_time = str()
```
Comment (rahullepakshi) [inline tests/nvmeof/workflows/ha.py:478]: Seems like counters are not getting updated? 

2024-10-11 19:43:02,753 (cephci.test_ceph_nvmeof_loadbalancing) [INFO] - cephci.cephci.tests.nvmeof.workflows.ha.py:574 - [ ceph-nvme-8-2-mn0w9x-node10 ] Total time taken to scale up - 1280664.256041291 seconds
2024-10-11 19:43:02,753 (cephci.test_ceph_nvmeof_loadbalancing) [INFO] - cephci.cephci.tests.nvmeof.workflows.ha.py:856 - {
    "scale-up-start-time": "",
    "scale-up-end-time": "Fri Oct 11 19:43:02 2024",
    "scale-up-start-counter-time": 0.0,
    "scale-up-end-counter-time": 1280664.256041291
}

File: tests/nvmeof/workflows/ha.py [hunk 2]
```diff
@@ -446,6 +460,127 @@ def power_on_off(self, gateway, action, wait_for_active_state=None):
...
+            LOG.info(
+                f"[ {hostname} ] Total time taken to scale up - {end_counter - start_counter} seconds"
+            )
+            return {
+                "scale-up-start-time": start_time,
+                "scale-up-end-time": end_time,
+                "scale-up-start-counter-time": start_counter,
+                "scale-up-end-counter-time": end_counter,
+            }
+
```
Comment (rahullepakshi) [inline tests/nvmeof/workflows/ha.py:583]: Post scale up , we need to do below

1. Add listeners to newly added GWs
2. Create few namespaces so new namespaces will take ANA_GRP of new GWs
3. Disconnect-all and discover again and run IO from initiator for all namespaces especially keep eye on IO progression for newly added namespaces in logs

**Comment by HaruChebrolu** [comment]
> Can we avoid running "gw info" command so many times? It creates confusion while analyzing logs
> 
> Also namespace list or probably any CLI output is logged twice. Any idea why? can we avoid it?

Can we avoid running "gw info" command so many times? It creates confusion while analyzing logs - "gw info" was running for each GW entity operation like "subsystem add", "listener add" for 2 reasons.
  - to check the GW health after evry operation
  - to use the few properties from "gw info" output to subsequent operations instead of hardcoding things
Also namespace list or probably any CLI output is logged twice. Any idea why? can we avoid it? - its being printed twice because of pretty_print and our own logger to check on exit_code. We need them inorder to diagnose on automation errors.

File: tests/nvmeof/workflows/ha.py (hunk 1 above)
Comment (HaruChebrolu) [reply tests/nvmeof/workflows/ha.py:478]: Done

File: tests/nvmeof/workflows/ha.py (hunk 2 above)
Comment (HaruChebrolu) [reply tests/nvmeof/workflows/ha.py:616]: I will take up this in next PR

**Review by rahullepakshi — APPROVED** [review]

**Review by Manohar-Murthy — APPROVED** [review]
lgtm.

**Pull request merged**